File: 1-request-lifecycle/1-java/src/main/java/academy/starci/requestlifecycle/ItemsController.java
package academy.starci.requestlifecycle;

import jakarta.validation.constraints.Min;
import java.util.List;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

/**
 * ItemsController handles the /items endpoint — request lifecycle pipeline demo.
 */
@RestController
@Validated
@RequestMapping("/items")
public class ItemsController {
    /**
     * service is the business layer providing item data.
     */
    private final ItemsService service;

    /**
     * Inject the service to keep the controller thin and testable.
     *
     * @param service the item business layer
     */
    public ItemsController(ItemsService service) {
        this.service = service;
    }

    /**
     * findAll handles GET /items to fetch the item list.
     *
     * @return the item list
     */
    @GetMapping
    public List<Item> findAll() {
        return service.findAll();
    }

    /**
     * findAllRestricted handles GET /items/restricted; RoleInterceptor blocks role != admin before the handler.
     *
     * @return the item list
     */
    @GetMapping("/restricted")
    public List<Item> findAllRestricted() {
        return service.findAll();
    }

    /**
     * findById handles GET /items/{id} with mandatory @Min(1) Bean Validation.
     *
     * @param id item id, must be a positive integer
     * @return the corresponding item
     */
    @GetMapping("/{id}")
    public Item findById(@PathVariable @Min(value = 1, message = "id must be a positive integer") long id) {
        return service.findById(id);
    }
}

File: 1-request-lifecycle/backend/2-csharp/ItemsController.cs
using System.ComponentModel.DataAnnotations;
using Microsoft.AspNetCore.Mvc;

namespace RequestLifecycle;

/// <summary>
/// ItemsController handles the /items endpoint — request lifecycle pipeline demo.
/// Controller is kept thin: no role check or timing logic here (those live in filters/middleware).
/// </summary>
[ApiController]
[Route("items")]
public class ItemsController : ControllerBase
{
    /// <summary>_service is the business layer providing item data.</summary>
    private readonly ItemsService _service;

    /// <summary>
    /// Inject the service to keep the controller thin and testable.
    /// </summary>
    /// <param name="service">The item business layer.</param>
    public ItemsController(ItemsService service) => _service = service;

    /// <summary>
    /// List handles GET /items to fetch the item list.
    /// </summary>
    /// <returns>The full item list.</returns>
    [HttpGet]
    public IEnumerable<Item> List() => _service.List();

    /// <summary>
    /// Restricted handles GET /items/restricted, guarded by RoleActionFilter.
    /// Only requests with role=admin reach this action; others get HTTP 403 from the filter.
    /// </summary>
    /// <returns>The full item list for authorized callers.</returns>
    [HttpGet("restricted")]
    [ServiceFilter(typeof(RoleActionFilter))]
    public IEnumerable<Item> Restricted() => _service.List();

    /// <summary>
    /// FindById handles GET /items/{id} with mandatory [Range] model validation.
    /// ModelState validation (400) fires before this action runs, proving the pipeline order.
    /// </summary>
    /// <param name="id">Item id, must be a positive integer.</param>
    /// <returns>The corresponding item.</returns>
    [HttpGet("{id}")]
    public Item FindById([Range(1, int.MaxValue, ErrorMessage = "id must be a positive integer")] int id)
        => _service.FindById(id);
}
